src/dbnode/storage/index.go

```diff
@@ -909,11 +1035,15 @@ func (i *nsIndex) Close() error {
 	if !i.isOpenWithRLock() {
 		return errDbIndexAlreadyClosed
 	}
+
 	i.state.closed = true
 
 	var multiErr xerrors.MultiError
 	multiErr = multiErr.Add(i.state.insertQueue.Stop())
 
+	// Wait for inflight queries to finish before closing blocks
+	i.queriesWg.Wait()
```

Comment:
Might be safer (or less likely to deadlock when changes are made) if you do this wait outside the lock. So:

```golang
i.state.Lock()
defer i.state.Unlock()

if !i.isOpenWithRLock() {
    return errDbIndexAlreadyClosed
}

// Signal our intent to close so that we stop accepting reads / writes.
i.state.closed = true

i.state.Unlock()

// Wait for in-flight queries to finish before continuing.
i.queriesWg.Wait()

// Reacquire the lock so we can complete the shutdown.
// Not even sure you need to re-acquire the lock at this point if you're willing
// to remove the defer at the top since technically no one else can call Close()
// after you mark is closed.
i.state.Lock()

var multiErr xerror.MultiError
multiErr = multiErr.Add(i.state.insertQueue.Stop())

...
```

Reply:
You understand the lifecycle better than I do, so take it or leave it